task/backend/executor/executor.go

```diff
@@ -303,38 +303,43 @@ func (p *asyncRunPromise) followQuery(wg *sync.WaitGroup) {
 	// Always need to call Done after query is finished.
 	defer p.q.Done()
 
-	select {
-	case <-p.ready:
-		// The promise was finished somewhere else, so we don't need to call p.finish.
-		// But we do need to cancel the flux. This could be a no-op.
-		p.q.Cancel()
-	case results, ok := <-p.q.Ready():
-		if !ok {
-			// Something went wrong with the flux. Set the error in the run result.
-			rr := &runResult{err: p.q.Err()}
-			p.finish(rr, nil)
+	var rwg sync.WaitGroup
+SelectLoop:
+	for {
+		select {
+		case <-p.ready:
+			// The promise was finished somewhere else, so we don't need to call p.finish.
+			// But we do need to cancel the flux. This could be a no-op.
+			p.q.Cancel()
 			return
+		case r, ok := <-p.q.Results():
+			if ok {
```

Comment:
Minor style nit: I would probably go with

```go
if !ok {
  break SelectLoop
}

rwg.Add(1)
// ...
```

just to reduce indentation in the `ok` path.